File: android/app/src/main/kotlin/dev/varuntej/aura/alarm/AlarmTones.kt
package dev.varuntej.aura.alarm

import android.content.Context
import android.content.res.AssetFileDescriptor
import android.media.RingtoneManager
import android.net.Uri
import android.util.Log

/**
 * The sounds an alarm can ring with, and the only place that decides which one.
 *
 * Mirrors `lib/core/constants/alarm_tones.dart` and
 * `backend/src/services/alarm_tones.py`. This is the copy that actually makes
 * noise at 3 AM, in a process with no Flutter engine and possibly no network,
 * which is why the table is duplicated here rather than read from Dart.
 *
 * Clips are bundled as FLUTTER assets, not `res/raw`. That is deliberate: the
 * settings picker previews `assets/alarm_tones/<slug>.wav` through just_audio
 * and this reads the same bytes out of the same APK entry, so the sound someone
 * chose and the sound that wakes them cannot drift apart.
 *
 * These clips do not declare beat metadata, so [AlarmRippleView] uses its
 * ambient timing while the audio loops.
 */
object AlarmTones {

    private const val TAG = "AuraAlarm"

    /** Buddy reads the reminder aloud over [BED_SLUG]. */
    const val BUDDY = "buddy"

    /** A sound from the user's own device, chosen with the system picker. */
    const val DEVICE = "device"

    /** What Buddy's voice rings over, and what an unknown slug degrades to. */
    const val BED_SLUG = "morning-clock-alarm"

    /** No beat this build knows about: the screen falls back to ambient ripples. */
    const val AMBIENT_BEAT_MS = 3200L

    data class Tone(val slug: String, val beatPeriodMs: Long = AMBIENT_BEAT_MS) {
        val assetPath: String get() = "flutter_assets/assets/alarm_tones/$slug.wav"
    }

    private val BUNDLED = listOf(
        Tone("morning-clock-alarm"),
        Tone("alert-alarm"),
        Tone("buzzer-alarm"),
        Tone("warning-buzzer"),
        Tone("street-public-alarm"),
        Tone("battleship-alarm"),
        Tone("retro-game-emergency"),
        Tone("rooster-crowing"),
        Tone("short-rooster-crowing"),
    ).associateBy { it.slug }

    /**
     * The clip a slug should play, or null when it is not a bundled tone.
     *
     * [BUDDY] resolves to the bed clip rather than to null: the spoken line is
     * layered on top by [AlarmService], and if the clip never arrives the bed is
     * what keeps the alarm from being silent.
     */
    fun bundled(slug: String): Tone? = when (slug) {
        BUDDY -> BUNDLED[BED_SLUG]
        else -> BUNDLED[slug]
    }

    /**
     * Milliseconds between ripples for a slug.
     *
     * Device-picked sounds and the system default have no BPM this process can
     * know, so they get the ambient period instead of a wrong one.
     */
    fun beatPeriodMs(slug: String): Long = bundled(slug)?.beatPeriodMs ?: AMBIENT_BEAT_MS

    /**
     * Open a bundled tone for playback, or null if it is not bundled or the APK
     * entry cannot be read.
     *
     * Returning null is a normal outcome, not an error: every caller falls
     * through to the device URI and then to the system default alarm sound.
     */
    fun openAsset(context: Context, slug: String): AssetFileDescriptor? {
        val tone = bundled(slug) ?: return null
        return try {
            context.assets.openFd(tone.assetPath)
        } catch (t: Throwable) {
            // Reachable if a build ships a slug whose asset was not bundled.
            // Loud, because it is silent-alarm territory and the fallback below
            // would otherwise hide it completely.
            Log.e(TAG, "alarm tone asset missing: ${tone.assetPath}", t)
            null
        }
    }

    /**
     * The URI picked with the system ringtone picker, with no fallback folded
     * into it. The service must know whether this exact source opened before it
     * records `device` as the ripple timing slug.
     */
    fun deviceUri(context: Context): Uri? {
        val stored = AlarmStore.deviceToneUri(context)
        if (stored.isNullOrBlank()) return null
        return runCatching { Uri.parse(stored) }.getOrNull()
    }

    /** Whether this slug is safe to persist as a global alarm preference. */
    fun isSelectable(slug: String): Boolean =
        slug.isEmpty() || slug == DEVICE || slug == BUDDY || BUNDLED.containsKey(slug)

    /** Exactly what the alarm played before tones existed. The last resort. */
    fun systemDefaultUri(): Uri? =
        RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM)
            ?: RingtoneManager.getDefaultUri(RingtoneManager.TYPE_RINGTONE)

    /**
     * The name of the user's picked sound, for the settings row. Blank if none.
     *
     * Resolved lazily rather than stored alongside the URI so a renamed or
     * deleted track shows the truth instead of a label kept from months ago.
     */
    fun deviceToneTitle(context: Context): String {
        val stored = AlarmStore.deviceToneUri(context)
        if (stored.isNullOrBlank()) return ""
        return try {
            RingtoneManager.getRingtone(context, Uri.parse(stored))?.getTitle(context).orEmpty()
        } catch (t: Throwable) {
            Log.w(TAG, "could not read the device tone title", t)
            ""
        }
    }
}

File: backend/src/services/alarm_tones.py
"""The sounds an alarm may ring with, and how a stored pick resolves.

Mirrors `lib/core/constants/alarm_tones.dart` and
`android/app/src/main/kotlin/dev/varuntej/aura/alarm/AlarmTones.kt`. The Kotlin
copy is the one that makes noise; this one exists so a slug that reaches the
device is always one the device can play.

Two stores, two meanings:

- ``users/{uid}.settings.alarm_tone`` is the user's default, set in the app.
- ``reminders/{id}.tone`` is a per-alarm override, which Buddy sets when someone
  asks for a specific one ("wake me at 6 with something gentle").

:func:`resolve_tone` collapses those to the single concrete slug that ships in
the sync row, so the device never has to know the precedence rule.
"""

from __future__ import annotations

# Bundled clips. Order mirrors the picker.
BUNDLED_TONES: tuple[str, ...] = (
    "morning-clock-alarm",
    "alert-alarm",
    "buzzer-alarm",
    "warning-buzzer",
    "street-public-alarm",
    "battleship-alarm",
    "retro-game-emergency",
    "rooster-crowing",
    "short-rooster-crowing",
)

# Buddy reads the reminder aloud in the user's chosen voice, over the bed clip.
# One of the tones rather than a flag beside them, so there is nothing to be off.
TONE_BUDDY = "buddy"

# A sound from the user's own device. The URI itself never leaves the phone; it
# is a path into their storage and means nothing here.
TONE_DEVICE = "device"

# Whatever the phone's default alarm sound is. Every account is on this today,
# and it is what an unresolvable pick degrades to.
TONE_SYSTEM_DEFAULT = ""

SELECTABLE_TONES: frozenset[str] = frozenset(
    BUNDLED_TONES + (TONE_BUDDY, TONE_DEVICE, TONE_SYSTEM_DEFAULT)
)

# What an LLM may choose. Neither special slug belongs here: "device" names a
# file only the phone can see, and the empty default is what omitting the
# argument already means.
ASSIGNABLE_TONES: tuple[str, ...] = BUNDLED_TONES + (TONE_BUDDY,)


def normalize_tone(value: object) -> str:
    """Coerce an untrusted stored value to a known slug, or to the default.

    firestore.rules lets the client write anything under ``users/{uid}``, and the
    tool argument is model-generated, so neither source can be trusted to hold a
    slug this build knows. An unrecognised value becomes the system default,
    which is the behaviour every alarm had before tones existed.
    """
    slug = str(value or "").strip().lower()
    return slug if slug in SELECTABLE_TONES else TONE_SYSTEM_DEFAULT


def resolve_tone(reminder_tone: object, user_default: object) -> str:
    """The one slug to send: per-alarm override, else user default, else system.

    Resolved here rather than on the device so the precedence rule lives in one
    place. The device receives a concrete answer and plays it.
    """
    override = normalize_tone(reminder_tone)
    if override:
        return override
    return normalize_tone(user_default)


def speaks_aloud(tone: str) -> bool:
    """Whether this tone needs a rendered wake-up clip cached before it rings."""
    return normalize_tone(tone) == TONE_BUDDY


async def user_default_tone(user_id: str) -> str:
    """The user's chosen default, or the system default on any failure.

    Never raises. This is read on the alarm sync path, and an alarm that rings
    with the phone's stock sound because a settings read blipped is a far better
    outcome than an alarm that does not ring at all.
    """
    import asyncio

    from .firebase import admin_firestore

    def _read() -> str:
        snapshot = admin_firestore().collection("users").document(user_id).get()
        if not snapshot.exists:
            return TONE_SYSTEM_DEFAULT
        settings = (snapshot.to_dict() or {}).get("settings") or {}
        return normalize_tone(settings.get("alarm_tone"))

    try:
        return await asyncio.to_thread(_read)
    except Exception as exc:
        from ..lib.logger import logger

        logger.warn("alarm_tones: default tone read failed; using system default", {
            "user_id": user_id,
            "error": str(exc),
            "error_type": type(exc).__name__,
        })
        return TONE_SYSTEM_DEFAULT
